File: 02.recursion_linkedList_tree/recursion.c

// recursion demo in c

#include <stdio.h>

int factorial(int n) {
    if (n == 0) {
        return 1;
    }
    return n * factorial(n - 1);
}


int recursion(int n){
    if (n == 0) {
        return 0; // base case
    }
    recursion(n - 1); // recursive call
}

int main() {
    int n = 5;
    recursion(n);
    return 0;
}

// linked list demo in c

#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

void printList(struct Node* n) {
    while (n != NULL) {
        printf("%d ", n->data);
        n = n->next;
    }
}

int main() {
    struct Node* head = NULL;
    struct Node* second = NULL;
    struct Node* third = NULL;

    head = (struct Node*)malloc(sizeof(struct Node));
    second = (struct Node*)malloc(sizeof(struct Node));
    third = (struct Node*)malloc(sizeof(struct Node));

    head->data = 1;
    head->next = second;

    second->data = 2;
    second->next = third;

    third->data = 3;
    third->next = NULL;

    printList(head);

    return 0;
}

File: 02.recursion_linkedList_tree/recursion.py
# recursion demo in python

def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n-1)


def fibonacci(n):
    if n < 2:
        return n
    else:
        return fibonacci(n-1) + fibonacci(n-2)

# 131231


def sum_of_digits(n):
    if n == 0:
        return 0
    else:
        return n % 10 + sum_of_digits(n//10)


# 10 20
def power(base, exp):
    if exp == 0:
        return 1
    else:
        return base * power(base, exp-1)


print(power(2, 3))  # 8
print(power(30, 40))  # 81
